### PR TITLE
drivers: spi: Add support for half-duplex (3-wire) SPI

### PR DESCRIPTION
**Pull request by ThreeEights**
Add support for half-duplex (3-wire) SPI operation using the Raspberry Pi Pico PIO.  To allow control of the size of the driver, including half-duplex support is optional, under the control of Kconfig options.

The original PIO source code is also included as a reference.

### CODE REVIEW

**Comment by ThreeEights** [comment]
Drat - the tx/rx counts in the 3-wire logic were wrong.  Corrected.

File: drivers/spi/spi_rpi_pico_pio.pio [hunk 1]
```diff
@@ -0,0 +1,78 @@
+; Copyright (c) 2023 Stephen Boylan <stephen.boylan@beechwoods.com>
```
Comment (soburi) [inline drivers/spi/spi_rpi_pico_pio.pio:1]: Is this file needed?

File: drivers/spi/Kconfig.rpi_pico [hunk 2]
```diff
@@ -9,3 +9,31 @@ config SPI_RPI_PICO_PIO
 	select PICOSDK_USE_CLAIM
 	help
 	  Enable driving SPI via PIO on the PICO
+
+if SPI_RPI_PICO_PIO
```
Comment (soburi) [inline drivers/spi/Kconfig.rpi_pico:13]: These config items should be better to create as dts properties.

Comment (ThreeEights) [reply drivers/spi/Kconfig.rpi_pico:13]: The Kconfig options determine whether or not some good-sized code chunks are included when compiling spi_rpi_pico_pio.c.  Since half-duplex and word size options are set via the run-time operation parameter, they can't be defined in DTS.  I may be over-fussy about saving program memory here, and I'd appreciate any comments on whether this detailed level of control is appropriate.

File: drivers/spi/spi_rpi_pico_pio.pio (hunk 1 above)
Comment (soburi) [reply drivers/spi/spi_rpi_pico_pio.pio:1]: I think these codes are unnecessary as it is already embedded in the C source as a comment.

File: drivers/spi/Kconfig.rpi_pico (hunk 2 above)
Comment (soburi) [reply drivers/spi/Kconfig.rpi_pico:13]: I'm concerned that pio-spi will be forced into a common configuration if there are instances on each of pio0 and pio1. There is no problem if these options are capabilities, but even in that case, I think they will be necessary to set them individually.

Comment (ThreeEights) [reply drivers/spi/Kconfig.rpi_pico:13]: I'm confused:  The Kconfig options don't control the configuration of the individual PIO peripherals;  that's done through DTS and the operation parameter.  The Kconfig options control what features are compiled into the driver.  If neither PIO requires half-duplex support, leaving CONFIG_SPI_RPI_PICO_PIO_HALF_DUPLEX unset will leave that code out of the compilation. If either PIO supports half-duplex, that option is necessary to make the half-duplex code available to whichever PIO needs it.  But it's still possible for one PIO to use a half-duplex configuration while the other uses the minimal 4-wire code.

Comment (soburi) [reply drivers/spi/Kconfig.rpi_pico:13]: I think these configs can be replaced with dts and `DT_ANY_INST_HAS_PROP_STATUS_OKAY` macro.
We can write "if any spi-pio node has ... property, then ...." with this macro. It has the same meaning at the logical level.
By doing this, you should be able to respond more flexibly to the "hardware" configuration of spi-pio.

And these configs seem to define "hardware capabilities".
So, it is better to define in dts.
(I think that it is more appropriate to consider the functions configured by pio as "hardware" in the zephyr driver structure.)

> But it's still possible for one PIO to use a half-duplex configuration (by defining sio-gpios in DTS) while the other uses the minimal 4-wire code.

This isn't easy to understand from the description.

Comment (ThreeEights) [reply drivers/spi/Kconfig.rpi_pico:13]: @soburi - Turns out you were right:  the HALF_DUPLEX config option can, indeed, be replaced with DT_ANY_INST_HAS_PROP_STATUS_OKAY, based on detecting the presence of the sio-gpios in the DTS.  I've updated the code to do so.  Thanks for pursuing that issue!

File: drivers/spi/spi_rpi_pico_pio.c [hunk 3]
```diff
@@ -258,28 +543,228 @@ static void spi_pico_pio_txrx(const struct device *dev)
...
+
+#if defined(CONFIG_SPI_RPI_PICO_PIO_8_BIT)
+			case 1: {
+				if (txbuf) {
+					txrx = ((uint8_t *)txbuf)[data->tx_count];
+				}
+				spi_pico_pio_sm_put8(data->pio, data->pio_sm, txrx);
+				data->tx_count++;
+			} break;
+#endif /*CONFIG_SPI_RPI_PICO_PIO_8_BIT*/
```
Comment (soburi) [inline drivers/spi/spi_rpi_pico_pio.c:575]: This is just a matter of style and preference, but there are ifdefs all over the place, so it would be nice to see them put together. For example, how about changing the function to something like this?

```
static inline int spi_pico_pio_sm_put(PIO pio, uint sm, void *dataptr, size_t bits)
{
        switch(bits) {
		case 4:
			*((io_rw_32 *)&pio->txf[sm]) = *(uint32_t*)dataptr;
			return 4;
		case 2:
			*((io_rw_8 *)&pio->txf[sm]) = *(uint16_t*)dataptr;
			return 2;
		case 1:
			*((io_rw_8 *)&pio->txf[sm]) = *(uint8_t*)dataptr;
			return 1;
		default:
			return 0;
        }
}
```

File: drivers/spi/spi_rpi_pico_pio.c [hunk 4]
```diff
@@ -41,44 +44,157 @@ struct spi_pico_pio_data {
...
 	uint32_t dfs;
 };
 
-RPI_PICO_PIO_DEFINE_PROGRAM(spi_cpol_0_cpha_0, 0, 1,
+/* ------------ */
+/* spi_mode_0_0 */
+/* ------------ */
+
+#define SPI_MODE_0_0_WRAP_TARGET 0
+#define SPI_MODE_0_0_WRAP        1
```
Comment (soburi) [inline drivers/spi/spi_rpi_pico_pio.c:61]: These lines are bit verbose.
These only use one time, I think no need to define.

File: drivers/spi/Kconfig.rpi_pico [hunk 5]
```diff
@@ -9,3 +9,26 @@ config SPI_RPI_PICO_PIO
...
+	bool "PIO 16-bit support"
+	help
+	  Enable any addtional code needed to
+	  support 16-bit transfers.
+
+config SPI_RPI_PICO_PIO_32_BIT
+	bool "PIO 32-bit support"
+	help
+	  Enable any addtional code needed to
+	  support 32-bit transfers.
```
Comment (soburi) [inline drivers/spi/Kconfig.rpi_pico:32]: I think the same consideration as for HALF_DUPLEX is required here as well.

File: drivers/spi/spi_rpi_pico_pio.c (hunk 3 above)
Comment (tbursztyka) [reply drivers/spi/spi_rpi_pico_pio.c:575]: Unless one wants to save some ROM bytes, in this case the defines are useful. (unless, again, these defines disappear and become DTS properties)

Comment (ThreeEights) [reply drivers/spi/spi_rpi_pico_pio.c:575]: Condensing the ..._put and ..._get code, as @soburi suggested, is a good idea.  That makes the source code a lot shorter and cleaner.  I'll try to get that to work.

File: drivers/spi/spi_rpi_pico_pio.c (hunk 4 above)
Comment (ThreeEights) [reply drivers/spi/spi_rpi_pico_pio.c:61]: I was trying to create a consistent pattern across all the PIO blocks that works with the current Zephyr PIO definitions.  I'd prefer to leave these.

File: drivers/spi/Kconfig.rpi_pico (hunk 5 above)
Comment (ThreeEights) [reply drivers/spi/Kconfig.rpi_pico:32]: Alas, the bit size being transferred can be defined at either build time or at run time, and can be changed on the fly.  Since this is a software configuration topic rather than hardware, it seems to be this belongs in Kconfig.  Which brings me back to my earlier question:  Does it make sense to include this fine a level of control?  Are the savings worth the overhead in the source?

File: drivers/spi/spi_rpi_pico_pio.c (hunk 4 above)
Comment (soburi) [reply drivers/spi/spi_rpi_pico_pio.c:61]: I got it. Let's do as so.

File: drivers/spi/Kconfig.rpi_pico (hunk 5 above)
Comment (tbursztyka) [reply drivers/spi/Kconfig.rpi_pico:32]: nah indeed, you can get rid of such config options. Just build the driver with the 3 size support.

Comment (ThreeEights) [reply drivers/spi/Kconfig.rpi_pico]: @soburi @tbursztyka - Good enough.  Have a gander at the latest force push and see what you think (assuming all the checks pass ...).

**Comment by ThreeEights** [comment]
@soburi, @tbursztyka, @yonsch - Can you please take a look at the latest changes and see if the issues are resolved?  Thanks!

**Comment by Ablu** [comment]
@ThreeEights: Is this the last gap that is needed to use bluetooth/wifi on the Pico W? Or is there further glue code between the CYW43xxx driver and this required?

Do you happen to have device-trees for enabling the cyw43439 around? Then I would be happy to give it a try :).

**Comment by ThreeEights** [comment]
@Ablu - Alas, no:  The current Infineon WiFi driver for the CYW43439 does not have SPI support.  If you're interested, you can view my latest attempt (as of yesterday!) on my fork:
https://github.com/ThreeEights/zephyr on branch airoc-wifi-spi

File: drivers/spi/spi_rpi_pico_pio.pio (hunk 1 above)
Comment (soburi) [reply drivers/spi/spi_rpi_pico_pio.pio:1]: @ThreeEights 
Please remove this file.

Comment (ThreeEights) [reply drivers/spi/spi_rpi_pico_pio.pio]: Removed.

**Review by soburi — APPROVED** [review]

**Comment by soburi** [comment]
@tbursztyka 
May we have your opinion?

**Review by tbursztyka — APPROVED** [review]

**Comment by ThreeEights** [comment]
@yonsch - Could you please take a look at this pull request?  Thanks.
